### ISSUE TITLE
Incorrect call not spotted using Union and TypeVar.

### ISSUE THREAD

**Issue by JulienPalard** [python/mypy#12450]
Playing around with:

```python
from typing import TypeVar, Union

T = TypeVar("T")


def f(x: Union[T, int], y: T) -> None:
    pass


f("a", 1)
```

the `f` call on the last line seems invalid to me:
 - `x` is of type `Union[T, int]` and is given a `str`, so `T` is `str`.
 - `y` is of type `T` and is given an `int` so `T` is `int`.

For me here `T` cannot be decided and mypy should probably whine here? Maybe mypy could tell:

    error: Invalid arguments for function "f": "T" cannot be "str" and "int" at the same time.

Currently it looks like `int` and `str` are joined, giving `object`:

```python
def f(x: Union[T, int], y: T) -> T:
    pass


reveal_type(f("a", 1))
```

gives:

    test.py:10: note: Revealed type is "builtins.object*"

**Comment by erictraut** [comment]
Mypy's behavior is correct here by the rules of PEP 484. It's "solving" for type `T`. There are multiple valid solutions that match all of the constraints. One valid solution is `object`, which is what mypy evaluates. Another valid solution is `str | int`, which is what pyright evaluates. In any case, no error should be reported because a type exists that satisfy all of the constraints for `T`.

If you want to constrain `T` to a particular subset of types, you could use a constrained TypeVar instead. 
```python
T = TypeVar("T", int, str)
```
If you make this replacement in your code sample above, it should generate an error, but mypy still accepts it. (Pyright rejects it in this case because there's no solution that satisfies all of the constraints.) So I think there's a bug in mypy in this case, but it's not the bug that was originally reported.

**Comment by JulienPalard** [comment]
Oooohhh, ok, it works to make:
```python
from typing import TypeVar, Union

T = TypeVar("T")


def f(x: Union[T, int], y: T) -> T:
    pass

reveal_type(f({1, 2}, [3, 4]))
```

work by choosing `typing.Collection*[builtins.int]`, make sense.

**Comment by erictraut** [comment]
Oh yes, my mistake. Mypy does generate an error in that case, as I would expect.
